File: week1-c/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <string.h>

string validateCard(long number);
string checkSum(long ccnumber);

int main(void)
{

    // Get CC number
    long ccnumber;
    do
    {
        ccnumber = get_long("Please enter the credit card number: ");
    }
    while (ccnumber < 0);

    string validity = "VALID";
    string invalidity = "INVALID";
    string control_check = checkSum(ccnumber);

    if (strcmp(control_check, validity) == 0)
    {
        string validation = validateCard(ccnumber);
        printf("%s\n", validation);
    }
    else
    {
        printf("INVALID\n");
    }
}


// Validate if VALID card number
// If valid, state card type
// Paramaters:
// AMEX = 15-digit; MASTERCARD = 16-digit; VISA = 13- and 16-digit numbers
// AMEX starts with 34 or 37; MASTERCARD starts [51:55]; VISA starts 4-.
string validateCard(long number)
{

    if ((number / (100000000000000)) >= 51 && (number / (100000000000000)) <= 55)
    {
        return "MASTERCARD";
    }
    else if ((number / 1000000000000000) == 4 || (number / 1000000000000) == 4)
    {
        return "VISA";
    }
    else if ((number / (10000000000000)) == 34 || (number / (10000000000000)) == 37)
    {
        return "AMEX";
    }
    else
    {
        return "INVALID";
    }

}


// Create a checksum function to get the digits and perform the checksum
string checkSum(long ccnumber)
{
    // Loop to obtain the last digit of the creditcard number, dividing by 10 to reduce the numner by one decimal place,
    // and then obtaining the last digit again in the updated loop cycle
    int last_digit;
    long number = ccnumber;
    for (int i = 1; i < 17; i++)
    {
        last_digit = number % 10;
        number = (number / 10);
    }

    // Variables used in calculations:
    // Rules being followed
    // Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together.
    // Add the sum to the sum of the digits that weren’t multiplied by 2.
    // If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!

    int times_one;
    int sum_double = 0;
    int sum_single = 0;
    int times_two;
    long number_double = ccnumber;
    int last_digit_double;

    // Looping 1 to 17 to cover 16 elements, initializing i to 1 because the nested conditional is doing an even odd test.
    for (int i = 1; i < 17; i++)
    {
        last_digit_double = number_double % 10;
        number_double = (number_double / 10);

        // If odd, collect last digit, else, collect the (last digit * 2)
        // Setting this up to double every other number in the CC numner starting with the penultimate numner
        // When initialized to 1, the pensultimate would begin at i = 2 - an even number
        if (i % 2 != 0)
        {
            times_two = last_digit_double;
            sum_double += 0;
            sum_single += times_two;
        }
        else
        {

            // The conditional below is to determine if one of the doubled values is a two digit number, split that number into
            // two separate digits and then add the digits. Ex - doubling 7 would give 14, the conditional would split 14 to 1 and 4,
            // summing them to equal 5.
            times_two = last_digit_double * 2;
            if (times_two >= 10)
            {
                times_two = ((last_digit_double * 2) % 10) + floor((last_digit_double * 2) / 10);
            }
            else
            {
                times_two = times_two + 0;
            }

            sum_double += times_two;
            sum_single += 0;
        }
    }

    // String congrol is going to hold either "VALID" or "INVALID"
    // The string value is being returned to an if else condition written in main(). In short, if the rules of the program are satisfied
    // by the calculations above and immediately below, the program will return "VALID" to that conditional and then determine which
    // card type the number corresponds to.
    string control;
    int total = sum_single + sum_double;


    // Conditional stating if remainder when divided by 10 is 0, do this else. Remainder of 0 is the solution to the problem statement
    // described above
    if (total % 10 == 0)
    {
        control = "VALID";
    }
    else
    {
        control = "INVALID";
    }
    return control;
}

File: week6-python/sentimental-credit/credit.py
from cs50 import get_string
import sys
import re
import math

''' Implement a program that determines whether a provided credit card number is valid according to Luhn's Algorithm:
    AE start 34 or 37, 15-digit number
    MC start with 51-55, 16-digit number
    Visa start with a 4, 13- or 16-digit number

    Luhn's Algorithm
    - Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together.
    - Add the sum to the sum of the digits that weren’t multiplied by 2.
    - If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
'''

# Defining a dictionary to store cc number, cc type
dict = {
    "n": "000000000000000",
    "l": 0
}

# Defining some strings for the Redex
visa_string = "^4"
amex_string = "^[34,37]{2}"
mc_string = "^[5]{1}[1-5]{1}"

dict["n"] = get_string("Number: ")
dict["l"] = len(dict["n"])


def credit_type(dict):

    if (dict["l"]) == 13:
        v_matching(dict)

    elif (dict["l"]) == 15:
        amex_matching(dict)

    elif (dict["l"]) == 16:
        if dict["n"][0] == "4":
            v_matching(dict)
        else:
            master_matching(dict)
    else:
        print("INVALID")


def luhn(dict):

    digits = []
    # Multiply every other digit by 2, then add digits together starting w penultimate
    m, mod, floor, c_int = 0, 0, 0, 0
    for char in dict["n"][len(dict["n"]) - 2::-2]:
        c_int = (2 * int(char))
        if c_int >= 10:
            mod = ((c_int) % 10)
            floor = math.floor(c_int / 10)
            digits.append(str(mod))
            digits.append(str(floor))
        else:
            digits.append(str(c_int))

    # Looping through digits list and summing each digit to the variable m
    for i in range(len(digits)):
        m += int(digits[i])

    # Step 2: Add the sum of the digits that weren't multiplied by 2
    # Looping backwards through string at dict["n"] 2 steps at a time
    summ = 0
    for char in dict["n"][::-2]:
        summ += int(char)

    # Step 3, if %10 is 0, the number passes Luhn test
    if (m + summ) % 10 == 0:
        return True
    else:
        return False


def v_matching(dict):

    if (re.match(visa_string, dict["n"])):
        print("VISA")
    else:
        print("INVALID")


def amex_matching(dict):

    if (re.match(amex_string, dict["n"])):
        print("AMEX")
    else:
        print("INVALID")


def master_matching(dict):
    if (re.match(mc_string, dict["n"])):
        print("MASTERCARD")
    else:
        print("INVALID")


if luhn(dict):
    credit_type(dict)
else:
    print("INVALID")
